src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -0,0 +1,478 @@
+pub struct InputPortView {}
+impl ShapeViewDefinition for InputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        init_shape(shape, Direction::In);
+
+        Self {}
+    }
+}
+
+/// Shape view for output port.
+#[derive(Debug,Default,Clone,Copy)]
+pub struct OutputPortView {}
+impl ShapeViewDefinition for OutputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        init_shape(shape, Direction::Out);
+        Self {}
+    }
+}
+
+/// Helper trait that describes a `ShapeViewDefinition` with a port shape.
+pub trait PortShapeViewDefinition = ShapeViewDefinition<Shape=shape::Shape>;
+
+/// Port definition.
+///
+/// A port must always be instantiated as an `InputPort` or an `OutputPort`. This determines its
+/// actual shape.
+///
+/// Example
+/// -------
+/// ```
+/// use graph_editor::component::node::port::InputPort;
+/// use graph_editor::component::node::port::OutputPort;
+///
+/// let input_port  = InputPort::default();
+/// let output_port = OutputPort::default();
+/// ```
+#[derive(Debug,CloneRef)]
+#[allow(missing_docs)]
+pub struct Port<T:PortShapeViewDefinition> {
+    pub data : Rc<PortData<T>>
+}
+
+// Manually implement this due to https://github.com/rust-lang/rust/issues/26925
+impl<T:PortShapeViewDefinition> Clone for Port<T>{
+    fn clone(&self) -> Self {
+        let data = Rc::clone(&self.data);
+        Self {data}
+    }
+}
+
+/// Type that represents an input port.
+pub type InputPort = Port<InputPortView>;
+
+/// Type that represents an output port.
+pub type OutputPort = Port<OutputPortView>;
+
+/// Internal data of `Port.
+#[derive(Debug,Clone)]
+#[allow(missing_docs)]
+pub struct PortData<T:PortShapeViewDefinition> {
+    /// Height of the port.
+    height       : Cell<f32>,
+    /// Width of the port in degrees.
+    width        : Cell<Angle<Degrees>>,
+    /// Radius of the inner circle that the port is constructed around.
+    inner_radius : Cell<f32>,
+    /// Location of the port along the inner circle.
+    position     : Cell<Angle<Degrees>>,
+    /// Indicates whether this port is facing inwards or outwards.
+    direction    : Cell<Direction>,
+
+    pub view : Rc<component::ShapeView<T>>
+}
+
+impl<T:PortShapeViewDefinition> Port<T> {
+
+    /// Internal constructor based on a given specification.
+    fn new(spec:Specification) -> Self {
+        let logger = Logger::new("node");
+        let view   = Rc::new(component::ShapeView::new(&logger));
+
+        let data   = Rc::new(PortData{
+            height       : Cell::new(spec.height),
+            width        : Cell::new(spec.width),
+            inner_radius : Cell::new(spec.inner_radius),
+            position     : Cell::new(spec.position),
+            direction    : Cell::new(spec.direction),
+            view
+        });
+        Self {data} . init()
+    }
+
```

Comment:
This is important. We've been discussing it for long time already. `new` should **always** behave just as `default`. Otherwise it is very confusing. `Port::new()` should just create a port with default values which you can change later. Please try not to use "builder pattern here". Seriously, I think the `Specification` structure is not needed here.